Core/include/Acts/Propagator/ActorConcepts.hpp

```diff
@@ -0,0 +1,87 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <concepts>
+#include <type_traits>
+#include <utility>
+
+namespace Acts {
+
+template <typename actor_t>
+concept ActorHasResult = requires { typename actor_t::result_type; };
+
+template <typename actor_t, typename propagator_state_t, typename stepper_t,
+          typename navigator_t, typename... Args>
+concept ActorHasActWithoutResult = requires(
+    const actor_t& a, propagator_state_t& state, const stepper_t& stepper,
+    const navigator_t& navigator, Args&&... args) {
+  {
+    a.act(state, stepper, navigator, std::forward<Args>(args)...)
```

Comment:
SonarCloud has a good point; this should be `std::move` instead (also for 3 other occurrences in this file).

Reply:
Just for me to understand: Isn't `Args&&` a "forward reference"?

Reply:
Also isn't it kind of weird that it complains about that on the concept?

Reply:
I think it's only a universal reference if it's a function argument.

Reply:
I'll give a bit of a verbose description of this; you are probably already familiar with most of this but just for posterity, because I think it can be a bit of a confusing topic. Might also be helpful for other people.

So first we will need to really understand universal references. A universal reference looks like this:

```cpp
template<typename T>
void my_function(T && t);
```

Why is this a universal reference? Because it can take both lvalue references and rvalue references due to something known as the C++ reference collapsing rules. Let's take a look at what happens when we call `my_function` with an lvalue reference:

```cpp
std::vector<int> t;
my_function(t);
```

In this case, `T` will resolve to be `std::vector<int> &`, and so the type of `t` will be `std::vector<int> &&&`. The reference collapsing rules refine that `T&&&` collapses to `T&`, so `t` has the type `T&`. It's an lvalue reference inside `my_function`, and `my_function` is able to use it as such.

Now let's call the same function with an rvalue reference instead:

```cpp
std::vector<int> t;
my_function(std::move(t));
```

In this case, `T` resolves to `std::vector<int> &&` so `t` has type `std::vector<int> &&&&` which resolves to `std::vector<int> &&`. The magic of universal references is that one templated function can support both lvalue and rvalue references. But we need to be careful. Let's imagine we want to use that reference inside of `my_function`:

```cpp
template<typename T>
void my_function(T && t) {
    std::vector<int>(t);
}
```

In the case where `T` is `std::vector<int> &`, this works as normal. It is passed as an lvalue reference, so you get the expensive copy operation. But what if we pass an rvalue reference to `my_function`? Inside of `my_function`, `t` is an lvalue, so even though it was moved into `my_function`, we still have to use the expensive copy operation instead of the much more efficient move operation. So let's "fix" `my_function`:

```cpp
template<typename T>
void my_function(T && t) {
    std::vector<int>(std::move(t));
}
```

This now works well if `my_function` is passed an rvalue, as it will give us the efficient copy constructor. But we've now introduced a massive bug in the code! If we pass `my_function` an lvalue, we are moving a reference which might put the vector into an invalid state. Remember, the caller of `my_function` still has a reference to that vector which `my_function` has now inadvertently moved. This could lead to a segmentation fault or worse.

So this is where `std::forward` comes in. It will cast `t` to `T &&` if it was passed as an rvalue, and to `T &` if it was an lvalue. That allows us to write `my_function` properly:

```cpp
template<typename T>
void my_function(T && t) {
    std::vector<int>(std::forward<T>(t));
}
```

The magic of this is that if `my_function` is passed an lvalue reference, it is forwarded as an lvalue and we get the safe, slow copy constructor. If it is passed an rvalue, it will move `t` instead. We know that this is safe because `t` was originally moved into `my_function`! This is what is called perfect forwarding.

So, we know that using `std::move` instead of `std::forward` is very dangerous. But now let's consider the following code:

```cpp
template<typename T>
struct my_struct {
    void my_function(T && t);
};
```

This situation is fundamentally different, because `T` is determined on the struct level rather than on the function level. In other words, `T` is not a universal reference, because `T` is already fixed to be either `U`, `U &`, or `U &&`. You could invoke `my_struct::my_function` in whatever way you want, and `T` will always be the same, e.g.:

```cpp
my_struct<std::vector<int>&> s;
std::vector<int> v;
s.my_function(v); // valid, passed as lvalue reference.
s.my_function(std::move(v)); // valid, but still an lvalue reference, because `T` is bound to `s` and is not a template of the function call
```

In this case, `std::forward` makes no sense. I don't think it's _harmful_ per se, but it doesn't do anything.

And this is why SonarCloud is complaining. The code has a template type on the outside struct (or, in this case, the outside concept) so it is not a universal reference.

Reply:
The story for concepts is a bit more complicated, because when you add a `requires` clause with arguments, those arguments are never actually instantiated, so you can't actually have a `requires` clause with template arguments:

```
template<typename T>
requires(T && t) { ... };
```

This doesn't make sense, because you don't actually "call" this required clause anywhere, so there is never any template deduction. For concepts, you can only specify templates on the concept itself, and you have to explicitly provide those types. Again, no deduction is ever involved. So there is no real possibility of having universal references here.